### PR TITLE
R4R: Modify asset fee queries

### PR DESCRIPTION
**Pull request by secret2830**
#1591

### CODE REVIEW

File: app/v1/asset/fees.go [hunk 1]
```diff
@@ -17,24 +19,82 @@ const (
...
+	return feeHandler(ctx, k, owner, sdk.NewCoin(fee.Denom, actualFeeMin))
+}
+
+// GatewayTokenIssueFeeHandler performs fee handling for issuing gateway token
+func GatewayTokenIssueFeeHandler(ctx sdk.Context, k Keeper, owner sdk.AccAddress, symbol string, fee sdk.Coin) sdk.Error {
+	// get the actual fee
+	actualFee := getGatewayTokenIssueFee(ctx, k, symbol)
+
+	// convert to native token min denom
+	actualFeeMin := convertFeeToNativeTokenMin(actualFee)
```
Comment (chengwenxi) [inline app/v1/asset/fees.go]: alwasys convert coins in client.

File: app/v1/asset/fees.go [hunk 2]
```diff
@@ -59,12 +119,91 @@ func feeHandler(ctx sdk.Context, k Keeper, feeAcc sdk.AccAddress, fee sdk.Coin)
...
+	// get params
+	params := k.GetParamSet(ctx)
+	gatewayAssetFeeRatio := params.GatewayAssetFeeRatio
+
+	// compute the native token issurance fee and gateway token issurance fee
+	nativeTokenIssueFee := getTokenIssueFee(ctx, k, symbol)
+	gatewayTokenIssueFee := sdk.NewDecFromInt(nativeTokenIssueFee).Mul(gatewayAssetFeeRatio)
+
+	// error ignored
+	gwTokenIssueFeeF64, _ := strconv.ParseFloat(gatewayTokenIssueFee.String(), 64)
```
Comment (chengwenxi) [inline app/v1/asset/fees.go]: Repalce float64 by sdk.Dec

File: app/v1/asset/fees.go [hunk 3]
```diff
@@ -59,12 +119,91 @@ func feeHandler(ctx sdk.Context, k Keeper, feeAcc sdk.AccAddress, fee sdk.Coin)
...
+	// compute the native token mint fee and gateway token mint fee
+	nativeTokenMintFee := getTokenMintFee(ctx, k, symbol)
+	gatewayTokenMintFee := sdk.NewDecFromInt(nativeTokenMintFee).Mul(gatewayAssetFeeRatio)
+
+	// error ignored
+	gwTokenMintFeeF64, _ := strconv.ParseFloat(gatewayTokenMintFee.String(), 64)
+
+	// round fee
+	return convertFeeToInt(gwTokenMintFeeF64)
+}
```
Comment (chengwenxi) [inline app/v1/asset/fees.go]: Refactor `getGatewayCreateFee`, `getTokenIssueFee`, `getGatewayTokenIssueFee`, `getGatewayTokenMintFee` return sdk.Coin with denom=iris-atto.

File: app/v1/asset/fees.go [hunk 4]
```diff
@@ -17,24 +19,67 @@ const (
...
 func GatewayFeeHandler(ctx sdk.Context, k Keeper, owner sdk.AccAddress, moniker string, fee sdk.Coin) sdk.Error {
-	// get params
-	params := k.GetParamSet(ctx)
-	gatewayBaseFee := params.CreateGatewayBaseFee
+	// get the actual fee
+	actualFee := getGatewayCreateFee(ctx, k, moniker)
+
+	// check if the provided fee is enough
+	if fee.IsLT(actualFee) {
+		return ErrInsufficientFee(k.Codespace(), fmt.Sprintf("insufficient gateway create fee: expected %s, got %s", actualFee, fee))
```
Comment (haifengxi) [inline app/v1/asset/fees.go]: gateway creation fee

File: app/v1/asset/fees.go [hunk 5]
```diff
@@ -17,24 +19,67 @@ const (
...
+}
+
+// TokenIssueFeeHandler performs fee handling for issuing token
+func TokenIssueFeeHandler(ctx sdk.Context, k Keeper, owner sdk.AccAddress, symbol string, fee sdk.Coin) sdk.Error {
+	// get the actual fee
+	actualFee := getTokenIssueFee(ctx, k, symbol)
+
+	// check if the provided fee is enough
+	if fee.IsLT(actualFee) {
+		return ErrInsufficientFee(k.Codespace(), fmt.Sprintf("insufficient token issurance fee: expected %s, got %s", actualFee, fee))
```
Comment (haifengxi) [inline app/v1/asset/fees.go]: token issuance fee

File: app/v1/asset/fees.go [hunk 6]
```diff
@@ -17,24 +19,67 @@ const (
...
+}
+
+// TokenMintFeeHandler performs fee handling for minting token
+func TokenMintFeeHandler(ctx sdk.Context, k Keeper, owner sdk.AccAddress, symbol string, fee sdk.Coin) sdk.Error {
+	// get the actual fee
+	actualFee := getTokenMintFee(ctx, k, symbol)
+
+	// check if the provided fee is enough
+	if fee.IsLT(actualFee) {
+		return ErrInsufficientFee(k.Codespace(), fmt.Sprintf("insufficient token mint fee: expected %s, got %s", actualFee, fee))
```
Comment (haifengxi) [inline app/v1/asset/fees.go]: token minting fee

File: app/v1/asset/fees.go [hunk 7]
```diff
@@ -17,24 +19,67 @@ const (
...
+func GatewayTokenIssueFeeHandler(ctx sdk.Context, k Keeper, owner sdk.AccAddress, symbol string, fee sdk.Coin) sdk.Error {
+	// get the actual fee
+	actualFee := getGatewayTokenIssueFee(ctx, k, symbol)
 
-	// check if the denom of fee is same as that of gatewayBaseFee
-	if fee.Denom != gatewayBaseFee.Denom {
-		return ErrIncorrectFeeDenom(k.Codespace(), fmt.Sprintf("incorrect fee denom: expected %s, got %s", gatewayBaseFee.Denom, fee.Denom))
+	// check if the provided fee is enough
+	if fee.IsLT(actualFee) {
+		return ErrInsufficientFee(k.Codespace(), fmt.Sprintf("insufficient gateway token issurance fee: expected %s, got %s", actualFee, fee))
```
Comment (haifengxi) [inline app/v1/asset/fees.go]: gateway token issuance fee

File: app/v1/asset/fees.go [hunk 8]
```diff
@@ -17,24 +19,67 @@ const (
...
+// GatewayTokenMintFeeHandler performs fee handling for minting gateway token
+func GatewayTokenMintFeeHandler(ctx sdk.Context, k Keeper, owner sdk.AccAddress, symbol string, fee sdk.Coin) sdk.Error {
+	// get the actual fee
+	actualFee := getGatewayTokenMintFee(ctx, k, symbol)
 
 	// check if the provided fee is enough
-	if fee.Amount.LT(actualFee) {
-		return ErrInsufficientFee(k.Codespace(), fmt.Sprintf("insufficient gateway create fee: expected %d, got %d", actualFee, fee.Amount))
+	if fee.IsLT(actualFee) {
+		return ErrInsufficientFee(k.Codespace(), fmt.Sprintf("insufficient gateway token mint fee: expected %s, got %s", actualFee, fee))
```
Comment (haifengxi) [inline app/v1/asset/fees.go]: gateway token minting fee

File: app/v1/asset/fees.go [hunk 9]
```diff
@@ -59,17 +104,80 @@ func feeHandler(ctx sdk.Context, k Keeper, feeAcc sdk.AccAddress, fee sdk.Coin)
...
+	params := k.GetParamSet(ctx)
+	gatewayBaseFee := params.CreateGatewayBaseFee
+
+	// compute the fee
+	fee := calcFeeByBase(moniker, gatewayBaseFee.Amount)
+
+	return sdk.NewCoin(sdk.NativeTokenMinDenom, convertFeeToInt(fee))
+}
+
+// getTokenIssueFee returns the token issurance fee
```
Comment (haifengxi) [inline app/v1/asset/fees.go]: // ... token issuance fee

File: app/v1/asset/fees.go [hunk 10]
```diff
@@ -59,17 +104,80 @@ func feeHandler(ctx sdk.Context, k Keeper, feeAcc sdk.AccAddress, fee sdk.Coin)
...
+func getTokenIssueFee(ctx sdk.Context, k Keeper, symbol string) sdk.Coin {
+	// get params
+	params := k.GetParamSet(ctx)
+	issueFTBaseFee := params.IssueTokenBaseFee
```
Comment (haifengxi) [inline app/v1/asset/fees.go]: issueFTBaseFee --> issueTokenBaseFee

File: app/v1/asset/fees.go [hunk 11]
```diff
@@ -59,17 +104,80 @@ func feeHandler(ctx sdk.Context, k Keeper, feeAcc sdk.AccAddress, fee sdk.Coin)
...
+	fee := calcFeeByBase(symbol, issueFTBaseFee.Amount)
+
+	return sdk.NewCoin(sdk.NativeTokenMinDenom, convertFeeToInt(fee))
+}
+
+// getTokenMintFee returns the token mint fee
+func getTokenMintFee(ctx sdk.Context, k Keeper, symbol string) sdk.Coin {
+	// get params
+	params := k.GetParamSet(ctx)
+	mintFTFeeRate := params.MintTokenFeeRatio
```
Comment (haifengxi) [inline app/v1/asset/fees.go]: mintFTFeeBase --> mintTokenFeeRatio

File: app/v1/asset/asset.go [hunk 12]
```diff
@@ -168,3 +169,37 @@ func GetKeyIDFromUniqueID(uniqueID string) string {
 		return strings.ToLower(fmt.Sprintf("i.%s", uniqueID))
 	}
 }
+
+// IsAssetIDValid checks if the given asset id is valid
+func IsAssetIDValid(id string) (bool, sdk.Error) {
```
[the remark this answers is not in the document]
Comment (secret2830) [reply app/v1/asset/asset.go]: Done

File: app/v1/asset/fees.go (hunk 2 above)
Comment (secret2830) [reply app/v1/asset/fees.go]: Done

File: app/v1/asset/fees.go (hunk 3 above)
Comment (secret2830) [reply app/v1/asset/fees.go]: Done

**Pull request merged**